File: backend/controllers/product_controller.go
package controllers
import (
	"net/http"
	"cronos-backend/middleware"
	"cronos-backend/models"
	"cronos-backend/services"
	"cronos-backend/utils"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)
type ProductController struct {
	ProductService *services.ProductService
}
func NewProductController(service *services.ProductService) *ProductController {
	return &ProductController{ProductService: service}
}
func (ctrl *ProductController) CreateProduct(c *gin.Context) {
	userID := middleware.GetUserID(c)
	var product models.Product
	if err := c.ShouldBindJSON(&product); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	if err := ctrl.ProductService.CreateProduct(userID, &product); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusCreated, "Product created", product)
}
func (ctrl *ProductController) GetMarketplaceProducts(c *gin.Context) {
	products, meta, err := ctrl.ProductService.GetMarketplaceProducts(
		c.Query("page"), c.Query("limit"),
		c.Query("shrimp_type"), c.Query("size"),
		c.Query("search"), c.Query("sort"),
		c.Query("min_price"), c.Query("max_price"),
	)
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponseWithMeta(c, http.StatusOK, "Products retrieved", products, meta)
}
func (ctrl *ProductController) GetProduct(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid product ID", nil)
		return
	}
	product, err := ctrl.ProductService.GetProduct(id)
	if err != nil {
		utils.NotFound(c, "Product not found")
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Product retrieved", product)
}
func (ctrl *ProductController) GetMyProducts(c *gin.Context) {
	userID := middleware.GetUserID(c)
	products, err := ctrl.ProductService.GetMyProducts(userID)
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "My products retrieved", products)
}
func (ctrl *ProductController) UpdateProduct(c *gin.Context) {
	userID := middleware.GetUserID(c)
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid product ID", nil)
		return
	}
	var req models.Product
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	product, err := ctrl.ProductService.UpdateProduct(userID, id, &req)
	if err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Product updated", product)
}
func (ctrl *ProductController) DeleteProduct(c *gin.Context) {
	userID := middleware.GetUserID(c)
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid product ID", nil)
		return
	}
	if err := ctrl.ProductService.DeleteProduct(userID, id); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Product deleted", nil)
}

File: backend-spring/src/main/java/com/cronos/controller/ProductController.java
package com.cronos.controller;

import com.cronos.dto.response.ApiResponse;
import com.cronos.entity.Product;
import com.cronos.service.ProductService;
import com.cronos.service.ReviewService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;
import java.util.UUID;

@RestController
@RequestMapping("/api/products")
@RequiredArgsConstructor
public class ProductController {
    private final ProductService productService;
    private final ReviewService reviewService;

    @GetMapping
    public ResponseEntity<?> marketplace(
            @RequestParam(required = false) String page, @RequestParam(required = false) String limit,
            @RequestParam(required = false) String shrimp_type, @RequestParam(required = false) String size,
            @RequestParam(required = false) String search, @RequestParam(required = false) String sort_by,
            @RequestParam(required = false) String min_price, @RequestParam(required = false) String max_price) {
        var result = productService.getMarketplaceProducts(page, limit, shrimp_type, size, search, sort_by, min_price, max_price);
        return ResponseEntity.ok(ApiResponse.success("Products retrieved", result.get("products"), result.get("meta")));
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getProduct(@PathVariable UUID id) {
        return ResponseEntity.ok(ApiResponse.success("Product retrieved", productService.getProduct(id)));
    }

    @GetMapping("/{id}/reviews")
    public ResponseEntity<?> getReviews(@PathVariable UUID id) {
        return ResponseEntity.ok(ApiResponse.success("Reviews retrieved", reviewService.getProductReviews(id)));
    }

    @GetMapping("/my")
    public ResponseEntity<?> getMyProducts(Authentication auth) {
        return ResponseEntity.ok(ApiResponse.success("Products retrieved", productService.getMyProducts((UUID) auth.getPrincipal())));
    }

    @PostMapping
    public ResponseEntity<?> create(Authentication auth, @RequestBody Product product) {
        return ResponseEntity.ok(ApiResponse.success("Product created", productService.createProduct((UUID) auth.getPrincipal(), product)));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> update(Authentication auth, @PathVariable UUID id, @RequestBody Product req) {
        try {
            return ResponseEntity.ok(ApiResponse.success("Product updated", productService.updateProduct((UUID) auth.getPrincipal(), id, req)));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @DeleteMapping("/{id}")
    public ResponseEntity<?> delete(Authentication auth, @PathVariable UUID id) {
        try {
            productService.deleteProduct((UUID) auth.getPrincipal(), id);
            return ResponseEntity.ok(ApiResponse.success("Product deleted", null));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }
}
